File: Java/MIMIC/src/ejemplosencillo/EDA/UMDA.java
package ejemplosencillo.EDA;

import ejemplosencillo.base.IFuncionObjetivo;
import java.util.Collections;
import java.util.LinkedList;

/**
 *
 * @author Marco Sotelo
 */
public class UMDA extends EDA{
    public UMDA(Integer Dimension, Integer Iteraciones, Integer Poblacion, Integer Seleccion, IFuncionObjetivo funcion) {
        super(Dimension, Iteraciones, Poblacion, Seleccion, funcion);
    }

    @Override
    public LinkedList<Float> pdf(LinkedList<Item> de) {
        LinkedList<Float> probabilidades = new LinkedList<>();
        for(int i=0; i<this.Dimension; i++){
            probabilidades.add(0.0f);
        }
        for (Item item: de){
            for (int i=0; i<item.getValues().size(); i++){
                probabilidades.set(i, probabilidades.get(i)+item.getValues().get(i));
            }
        }
        for(int i=0; i<this.Dimension; i++){
            probabilidades.set(i, probabilidades.get(i)/de.size());
        }
        return probabilidades;
    }

    @Override
    public LinkedList<Item> Muestrear(LinkedList<Float> pdf, Integer Num) {
        LinkedList<Item> items = new LinkedList<>();
        for (int i=0; i<Num; i++){
            items.add(new Item());
            items.getLast().Generar(pdf);
            items.getLast().Evaluar(this.funcion);
        }
        Collections.sort(items);
        return items;
    }

}

File: kotlin/MIMIC/src/ejemplosencillo/EDA/UMDA.kt
package ejemplosencillo.EDA

import ejemplosencillo.base.IFuncionObjetivo
import java.util.Collections
import java.util.LinkedList

/**
 *
 * @author Marco Sotelo
 */
class UMDA(Dimension: Integer, Iteraciones: Integer, Poblacion: Integer, Seleccion: Integer, funcion: IFuncionObjetivo) : EDA(Dimension, Iteraciones, Poblacion, Seleccion, funcion) {

    @Override
    fun pdf(de: LinkedList<Item>): LinkedList<Float> {
        val probabilidades = LinkedList()
        for (i in 0 until this.Dimension) {
            probabilidades.add(0.0f)
        }
        for (item in de) {
            for (i in 0 until item.getValues().size()) {
                probabilidades.set(i, probabilidades.get(i) + item.getValues().get(i))
            }
        }
        for (i in 0 until this.Dimension) {
            probabilidades.set(i, probabilidades.get(i) / de.size())
        }
        return probabilidades
    }

    @Override
    fun Muestrear(pdf: LinkedList<Float>, Num: Integer): LinkedList<Item> {
        val items = LinkedList()
        for (i in 0 until Num) {
            items.add(Item())
            items.getLast().Generar(pdf)
            items.getLast().Evaluar(this.funcion)
        }
        Collections.sort(items)
        return items
    }

}
